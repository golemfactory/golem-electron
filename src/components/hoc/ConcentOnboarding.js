import React from 'react';
import { Redirect } from "react-router-dom";

import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'


import * as Actions from '../../actions'
import ConcentToS from './../concent/ConcentToS'
import ConcentOnboarding from './../concent/Onboarding'

/**
 * { HIGH ORDER COMPONENT }
 * { Onboarding component will check if user using the application first time or not}
 *
 * @class      ConcentToSComponent(name)
 * @param      {React Component}        ComposedComponent   The composed component which is we waiting for; see router in Container/App.js
 * @return     {ConnectedComponent>}                        { Returning connected component (Redux) }
 */
export var ConcentOnboardingComponent = function(ComposedComponent) {

    const mapStateToProps = state => ({
        concentTerms: state.info.concentTerms,
        concentSwitch: state.concent.concentSwitch,
        isOnboadingActive: !state.concent.hasOnboardingShown,
        showConcentToS: !state.info.isConcentTermsAccepted
    })

    const mapDispatchToProps = dispatch => ({
        actions: bindActionCreators(Actions, dispatch)
    })

    class ConcentOnboardingHoC extends React.Component {


        constructor(props) {
            super(props);
            this.state = {
                isOnboadingActive: props.isOnboadingActive,
                showConcentToS: props.showConcentToS
            }
        }

        componentWillReceiveProps(nextProps) {
            if(nextProps.showConcentToS !== this.props.showConcentToS){
                this.setState({
                    showConcentToS: nextProps.showConcentToS
                })
            }
            if(nextProps.isOnboadingActive !== this.props.isOnboadingActive){
                this.setState({
                    isOnboadingActive: nextProps.isOnboadingActive
                })
            }
        }

        _handleOnboardingDone = () => {
            const {actions, showConcentToS} = this.props
            actions.setConcentOnboardingShown()
            if(!showConcentToS){
                actions.toggleConcent(true, true)
            }
        }


        render() {
            const {concentSwitch, concentTerms} = this.props
            const {showConcentToS, isOnboadingActive} = this.state

            if(isOnboadingActive && concentSwitch){
                return <ConcentOnboarding handleOnboardingDone={this._handleOnboardingDone}/>
            }
            if(showConcentToS && concentSwitch){
                return <ConcentToS {...this.props}/>
            }
            return (
                <div>
                    <ComposedComponent/>
                </div>
            )
        }
    }

    return connect(mapStateToProps, mapDispatchToProps)(ConcentOnboardingHoC);
}

