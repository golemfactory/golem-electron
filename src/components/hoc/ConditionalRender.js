import React, { Component } from "react";

export default class ConditionalRender extends Component {
  render() {
    const { children, showIf } = this.props;
    if (showIf) {
      const childrenWithProps = React.Children.map(children, child =>
        React.cloneElement(child, { key: child.type.name })
      );
      return childrenWithProps;
    }
    return null;
  }
}
