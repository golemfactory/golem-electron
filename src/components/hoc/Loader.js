import React from 'react'
import uuid from 'uuid/v4'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { Spring } from 'react-spring/renderprops.cjs'

import * as Actions from '../../actions'
import golem_logo from './../../assets/img/golem-black.svg'

/**
 * {Â Class for loading Component which we will see while real component loading. }
 *
 * @class      Loading (name)
 */
export class Loading extends React.Component {

    componentDidMount() {
        require('../animation')
    }

    /**
     * { runAnim function updates style of div with React-Spring }
     *
     * @param      {number}     rotate  Transform attribute 
     * @param      {number}     alpha   Opacity attribute
     * @return     {Object}     { style object to change style }
     */
    runAnim(rotate, alpha) {
        return {
            transform: `translate(-50%, ${rotate}%)`,
            opacity: `${alpha}`
        }
    }

    render() {
        let mainButtonRotation = {
            rotate: -24,
            alpha: 1,
        };
        let defaultMainButtonRotation = {
            rotate: 0,
            alpha: 0
        }

        return (
            <div className="content__loading">
            <canvas id="stage" className="canvas" width="1534" height="841" style={{
                backgroundColor: 'rgba(0, 0, 0, 0)'
            }}></canvas>
            <Spring 
                from={defaultMainButtonRotation} 
                to={mainButtonRotation} 
                delay={1000}
                config={{ tension: 10, friction: 50 }}>
                    {({rotate, alpha}) => <div className="title" style={this.runAnim(rotate, alpha)}>
                                <img src={golem_logo} />
                                <div className="text"> worldwide
                                    <br/> supercomputer
                                </div>
                                <div id="loading" className="loading">
                                    <span></span>
                                    <span></span>
                                    <span></span>
                                    <span></span>
                                </div>
                            </div>
            }
                </Spring>
            </div>
        )
    }
}

/**
 * { HIGH ORDER COMPONENT }
 * { Loading component will redirect user to the loading screen while related component loading. }
 *
 * @class      LoadingComponent(name)
 * @param      {React Component}        ComposedComponent   The composed component which is we waiting for; see router in Container/App.js
 * @param      {Array}                  loadingIdArray      The loading identifier array
 * @return     {Array{ConnectedComponent, Component}}                        { Returning connected component (Redux) and Component (for testing)}
 */
export var LoadingComponent = function(ComposedComponent, loadingIdArray) {


    const mapStateToProps = state => ({
        loader: state.loader
    })

    const mapDispatchToProps = dispatch => ({
        actions: bindActionCreators(Actions, dispatch)
    })

    class Loader extends React.Component {

        constructor(props) {
            super(props);
        }

        componentWillMount() {
            let loadingId = uuid()
            this.loadingId = loadingId
            this.loadingIdArray = loadingIdArray ? [loadingId, ...loadingIdArray] : [loadingId]
        }

        _handleStartLoading(loadingText) {
            this.props.actions.startLoading(this.loadingId, loadingText)
        }

        _handleEndLoading(loadId) {
            this.props.actions.endLoading(loadId ? loadId : this.loadingId)
            return false
        }

        render() {
            let {loader, actions} = this.props
            let {loadingId, loadingIdArray} = this
            let passToChild = {
                startLoading: this._handleStartLoading,
                endLoading: this._handleEndLoading,
                loadingIdArray,
                loadingId
            }
            let loadingObject
            loadingIdArray.forEach((id) => {
                let loadObj = loader && loader[id];
                if (loadObj && loadObj.isLoading) {
                    loadingObject = loadObj;
                }
            });
            let isLoading = false

            if (loadingObject && loadingObject.isLoading) {
                isLoading = true
            }

            let styleOne = !isLoading ? {
                display: 'none'
            } : {}
            let styleTwo = isLoading ? {
                display: 'none'
            } : {}

            return (
                <div>
          <div style={{
                    width: '100%', ...styleOne
                }}>
            <Loading loadingText={loadingObject ? loadingObject.text : "Loading..."}/>
          </div>
          <div style={{
                    width: '100%', ...styleTwo
                }}>

                <ComposedComponent {...this.props} {...passToChild}/>
          </div>
        </div>
            );
        }
    }

    return [connect(mapStateToProps, mapDispatchToProps)(Loader), Loader]
}




