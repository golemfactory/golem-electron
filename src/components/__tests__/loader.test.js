jest.unmock('../hoc/Loader')

import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import TestUtils from 'react-dom/test-utils'
import configureStore from 'redux-mock-store';
import { LoadingComponent, Loading, Loader } from '../hoc/Loader'
import sinon from 'sinon'


const mockStore = configureStore()

describe('LoadingComponent Wrapper', () => {
    var wrapper,
        MockMainComponent,
        instance,
        set,
        WrapperComponent;

    beforeEach(() => {
        MockMainComponent = function Fake() {
                return (<div>Fake component</div>);
            }

        WrapperComponent = LoadingComponent(MockMainComponent)[0];
        wrapper = shallow(<Provider store={ mockStore({})}>
                                <WrapperComponent />
                          </Provider>);
        instance = wrapper.instance()
    });

    it('renders the Main Component as the root element', () => {
        expect(wrapper.find(WrapperComponent).length).toBe(1)
    });
});

describe('Loader', () => {
    let MockMainComponent;
    let Loader;

    beforeEach(() => {
        MockMainComponent = function Fake() {
                return (<div>Fake component</div>);
            }

        Loader = LoadingComponent(MockMainComponent, ['BLENDER_LOADER'])[1]
    });

    const setup = () => {
        const loader = {
            BLENDER_LOADER: {
                "isLoading": true,
                "text": "I am loading!"
            }
        }

        const wrapper = shallow(<Loader loader={loader}/>)
        return {
            wrapper
        }
    }

    const actions = {
        startLoading: jest.fn(),
        endLoading: jest.fn(),
    }

    it('should render an Loading container', () => {
        const {wrapper} = setup()
    });

    it('should call actions.startLoading', () => {
        const mockLoader = TestUtils.renderIntoDocument(<Loader actions={actions}/>);
        const loadingId = [123]
        mockLoader._handleStartLoading(loadingId)
        expect(actions.startLoading).toBeCalled()
    });

    it('should call actions.endLoading', () => {
        const mockLoader = TestUtils.renderIntoDocument(<Loader actions={actions}/>);
        const loadingId = [123]
        mockLoader._handleEndLoading(loadingId)
        expect(actions.endLoading).toBeCalled()
    });
    
    it('should call componentWillMount', () => {
        sinon.spy(Loader.prototype, 'componentWillMount');
        expect(Loader.prototype.componentWillMount.calledOnce).toBe(false)
        const wrapper = mount(<Loader />);
        expect(Loader.prototype.componentWillMount.calledOnce).toBe(true)
    })
})

describe('Loading', () => {

    const setup = () => {
        const wrapper = shallow(<Loading/>)

        return {
            wrapper
        }
    }

    const actions = {
        startLoading: jest.fn(),
        endLoading: jest.fn(),
    }

    it('should render an Loading container', () => {
        const {wrapper} = setup()
        expect(wrapper).toMatchSnapshot()
    });

    it('should render a stage canvas for animation', () => {
        const {wrapper} = setup()
        expect(wrapper.find('#stage').length).toBe(1)
    });


    it('should return style with runAnim', () => {
        const style = {
            transform: "translate(-50%, 50%)",
            opacity: "0.6"
        }
        let mockLoading = TestUtils.renderIntoDocument(<Loading />);
        expect(mockLoading.runAnim(50, 0.6).toString()).toBe(style.toString())
    });

    it('should call componentDidMount', () => {
        sinon.spy(Loading.prototype, 'componentDidMount');
        expect(Loading.prototype.componentDidMount.calledOnce).toBe(false)
        const wrapper = mount(<Loading />);
        expect(Loading.prototype.componentDidMount.calledOnce).toBe(true)
    })
});

