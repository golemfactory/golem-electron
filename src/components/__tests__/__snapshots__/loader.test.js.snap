// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading should render an Loading container 1`] = `
<div
  className="content__loading"
>
  <canvas
    className="canvas"
    height="841"
    id="stage"
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0)",
      }
    }
    width="1534"
  />
  <Spring
    config={
      Object {
        "friction": 50,
        "tension": 10,
      }
    }
    delay={1000}
    force={false}
    from={
      Object {
        "alpha": 0,
        "rotate": 0,
      }
    }
    immediate={false}
    inject={[Function]}
    native={false}
    reset={false}
    to={
      Object {
        "alpha": 1,
        "rotate": -24,
      }
    }
  >
    <Component />
  </Spring>
</div>
`;
